In [13]:
import altair as alt
import numpy as np
import pandas as pd

In [14]:
players_data = pd.read_csv("dsci-100-student-python/players.csv")
sessions_data = pd.read_csv("dsci-100-student-python/sessions.csv")

In [15]:
players_data

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  individualId  \
0            30.3     Morgan               Male    9           NaN   
1             3.8  Christian               Male   17           NaN   
2             0.0      Blake               Male   17           NaN   
3             0.7      Flora             Female   21           NaN   
4             0.1      Kylie               Male   21           NaN   
..            ...        ...                ...  ...           ...   
191           0.0     Bailey             Female   17           NaN   
192           0.3     Pascal               Male   22           NaN   
193           0.0      Dylan  Prefer not to say   17           NaN   
194           2.3     Harlow               Male   17           NaN   
195           0.2      Ahmed              Other   91           NaN   

     organizationName  
0                 NaN  
1                 NaN  
2                 NaN  
3                 NaN  
4                 NaN  
..                ...  
191               NaN  
192               NaN  
193               NaN  
194               NaN  
195               NaN  

[196 rows x 9 columns]

In [16]:
sessions_data

hashedEmail        start_time  \
0     bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...  30/06/2024 18:12   
1     36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  17/06/2024 23:33   
2     f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3...  25/07/2024 17:34   
3     bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...  25/07/2024 03:22   
4     36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  25/05/2024 16:01   
...                                                 ...               ...   
1530  36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  10/05/2024 23:01   
1531  7a4686586d290c67179275c7c3dfb4ea02f4d317d9ee0e...  01/07/2024 04:08   
1532  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...  28/07/2024 15:36   
1533  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...  25/07/2024 06:15   
1534  36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  20/05/2024 02:26   

              end_time  original_start_time  original_end_time  
0     30/06/2024 18:24         1.719770e+12       1.719770e+12  
1     17/06/2024 23:46         1.718670e+12       1.718670e+12  
2     25/07/2024 17:57         1.721930e+12       1.721930e+12  
3     25/07/2024 03:58         1.721880e+12       1.721880e+12  
4     25/05/2024 16:12         1.716650e+12       1.716650e+12  
...                ...                  ...                ...  
1530  10/05/2024 23:07         1.715380e+12       1.715380e+12  
1531  01/07/2024 04:19         1.719810e+12       1.719810e+12  
1532  28/07/2024 15:57         1.722180e+12       1.722180e+12  
1533  25/07/2024 06:22         1.721890e+12       1.721890e+12  
1534  20/05/2024 02:45         1.716170e+12       1.716170e+12  

[1535 rows x 5 columns]

Question 1

We are given two datasets, one named players and the other named sessions. The players dataset has 9 variables with 196 observations, whilst the sessions dataset has 5 variables with 1535 observations.

Players Dataset Variables:
- experience: User's experience with Minecraft 
- subscribe: Email subscription (Not necessary to use)
- hashEmail: Code used to identify specific users (Not necessary to use)
- played_hours: User's playing time.
- name: User's fake name
- gender: User's gender
- age: User's age
- individualid: User's individual ID (Not necessary to use)
- organizationName: User's organization name (Not necessary to use)

Sessions Dataset Variables:
- hashedEmail: Code used to identify specific users (Not necessary to use)
- start_time: When the user starts playing.
- end_time: When the user stops playing.
- original_start_time: When the user starts playing.
- original_end_time: When the user stops playing.

Question 2

The question that my group and I will be answering is question 2: "We are interested in demand forecasting, namely, what time windows are most likely to have large number of simultaneous players. This is because we need to ensure that the number of licenses on hand is sufficiently large to accommodate all parallel players with high probability."

The response variable is "start_time" and the explanatory variable would be "end_time", and "played_hours". By having "start_time" as our response variable, we are able to see when users first log on. When we first group by start_time, we would be able to see what time a good majority of players log on. Having end_time and played_hours as our explanatory variables, we will be able to create a time window to see when there are most users during a certain time.

In [21]:
# Question 3
players_data.drop(columns = ["individualId", "organizationName"])

experience  subscribe                                        hashedEmail  \
0          Pro       True  f6daba428a5e19a3d47574858c13550499be23603422e6...   
1      Veteran       True  f3c813577c458ba0dfef80996f8f32c93b6e8af1fa9397...   
2      Veteran      False  b674dd7ee0d24096d1c019615ce4d12b20fcbff12d79d3...   
3      Amateur       True  23fe711e0e3b77f1da7aa221ab1192afe21648d47d2b4f...   
4      Regular       True  7dc01f10bf20671ecfccdac23812b1b415acd42c2147cb...   
..         ...        ...                                                ...   
191    Amateur       True  b6e9e593b9ec51c5e335457341c324c34a2239531e1890...   
192    Veteran      False  71453e425f07d10da4fa2b349c83e73ccdf0fb3312f778...   
193    Amateur      False  d572f391d452b76ea2d7e5e53a3d38bfd7499c7399db29...   
194    Amateur      False  f19e136ddde68f365afc860c725ccff54307dedd13968e...   
195        Pro       True  d9473710057f7d42f36570f0be83817a4eea614029ff90...   

     played_hours       name             gender  age  
0            30.3     Morgan               Male    9  
1             3.8  Christian               Male   17  
2             0.0      Blake               Male   17  
3             0.7      Flora             Female   21  
4             0.1      Kylie               Male   21  
..            ...        ...                ...  ...  
191           0.0     Bailey             Female   17  
192           0.3     Pascal               Male   22  
193           0.0      Dylan  Prefer not to say   17  
194           2.3     Harlow               Male   17  
195           0.2      Ahmed              Other   91  

[196 rows x 7 columns]

In [24]:
# Question 3
sessions_data

hashedEmail        start_time  \
0     bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...  30/06/2024 18:12   
1     36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  17/06/2024 23:33   
2     f8f5477f5a2e53616ae37421b1c660b971192bd8ff77e3...  25/07/2024 17:34   
3     bfce39c89d6549f2bb94d8064d3ce69dc3d7e72b38f431...  25/07/2024 03:22   
4     36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  25/05/2024 16:01   
...                                                 ...               ...   
1530  36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  10/05/2024 23:01   
1531  7a4686586d290c67179275c7c3dfb4ea02f4d317d9ee0e...  01/07/2024 04:08   
1532  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...  28/07/2024 15:36   
1533  fd6563a4e0f6f4273580e5fedbd8dda64990447aea5a33...  25/07/2024 06:15   
1534  36d9cbb4c6bc0c1a6911436d2da0d09ec625e43e6552f5...  20/05/2024 02:26   

              end_time  original_start_time  original_end_time  
0     30/06/2024 18:24         1.719770e+12       1.719770e+12  
1     17/06/2024 23:46         1.718670e+12       1.718670e+12  
2     25/07/2024 17:57         1.721930e+12       1.721930e+12  
3     25/07/2024 03:58         1.721880e+12       1.721880e+12  
4     25/05/2024 16:12         1.716650e+12       1.716650e+12  
...                ...                  ...                ...  
1530  10/05/2024 23:07         1.715380e+12       1.715380e+12  
1531  01/07/2024 04:19         1.719810e+12       1.719810e+12  
1532  28/07/2024 15:57         1.722180e+12       1.722180e+12  
1533  25/07/2024 06:22         1.721890e+12       1.721890e+12  
1534  20/05/2024 02:45         1.716170e+12       1.716170e+12  

[1535 rows x 5 columns]

Question 4

I believe that KNN Regression would be a good method and model to use for this dataset. Because there are some players who play for a few minutes, and some players who could possibly play for hours, we need to be as accurate as possible, thus we must account for every small and or large time session played. A possible weakness with using this model is that outliers can heavily skew the data once finished, concluding in an inaccurate representation of the data. We are going to process the model by predicting it by using a 70/30 split. This is because the dataset that we want to use, which is the sessions dataset, has many observations, we believe that the result should be rather accurate. Because the dataset is rather large, I don't think cross validation is necessary. As previously mentioned the dataset is large enough, adding cross validation would make the process longer and only specific to this dataset which wouldn't be reusable in the future.